--- ./vanilla//arch/arm/include/asm/vfp.h	2010-02-23 17:38:51.000000000 +0200
+++ ./default//arch/arm/include/asm/vfp.h	2012-05-23 10:43:53.871935675 +0300
@@ -82,3 +82,8 @@
 #define VFPOPDESC_UNUSED_BIT	(24)
 #define VFPOPDESC_UNUSED_MASK	(0xFF << VFPOPDESC_UNUSED_BIT)
 #define VFPOPDESC_OPDESC_MASK	(~(VFPOPDESC_LENGTH_MASK | VFPOPDESC_UNUSED_MASK))
+
+#ifndef __ASSEMBLY__
+int vfp_flush_context(void);
+void vfp_reinit(void);
+#endif
