--- ./vanilla//drivers/gpu/drm/radeon/radeon_drv.c	2010-02-23 17:38:51.000000000 +0200
+++ ./default//drivers/gpu/drm/radeon/radeon_drv.c	2012-05-23 10:43:57.675935507 +0300
@@ -215,7 +215,7 @@
 static int __devinit
 radeon_pci_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
 {
-	return drm_get_dev(pdev, ent, &kms_driver);
+	return drm_get_pci_dev(pdev, ent, &kms_driver);
 }
 
 static void
