--- ./vanilla//include/net/bluetooth/l2cap.h	2010-02-23 17:38:51.000000000 +0200
+++ ./default//include/net/bluetooth/l2cap.h	2012-05-23 10:44:04.665935195 +0300
@@ -70,6 +70,7 @@
 #define L2CAP_LM_TRUSTED	0x0008
 #define L2CAP_LM_RELIABLE	0x0010
 #define L2CAP_LM_SECURE		0x0020
+#define L2CAP_LM_FLUSHABLE	0x0040
 
 /* L2CAP command codes */
 #define L2CAP_COMMAND_REJ	0x01
@@ -316,6 +317,7 @@
 	__u8		sec_level;
 	__u8		role_switch;
 	__u8		force_reliable;
+	__u8		flushable;
 
 	__u8		conf_req[64];
 	__u8		conf_len;
