--- ./vanilla//include/linux/completion.h	2010-02-23 17:38:51.000000000 +0200
+++ ./default//include/linux/completion.h	2012-05-23 10:44:04.548935199 +0300
@@ -77,6 +77,7 @@
 }
 
 extern void wait_for_completion(struct completion *);
+extern void wait_for_completion_io(struct completion *);
 extern int wait_for_completion_interruptible(struct completion *x);
 extern int wait_for_completion_killable(struct completion *x);
 extern unsigned long wait_for_completion_timeout(struct completion *x,
