--- ./vanilla//include/linux/netdevice.h	2010-02-23 17:38:51.000000000 +0200
+++ ./default//include/linux/netdevice.h	2012-05-23 10:44:04.601935198 +0300
@@ -113,9 +113,9 @@
     !defined(CONFIG_NET_IPGRE) &&  !defined(CONFIG_NET_IPGRE_MODULE) && \
     !defined(CONFIG_IPV6_SIT) && !defined(CONFIG_IPV6_SIT_MODULE) && \
     !defined(CONFIG_IPV6_TUNNEL) && !defined(CONFIG_IPV6_TUNNEL_MODULE)
-#define MAX_HEADER LL_MAX_HEADER
+#define MAX_HEADER (LL_MAX_HEADER + 20)
 #else
-#define MAX_HEADER (LL_MAX_HEADER + 48)
+#define MAX_HEADER (LL_MAX_HEADER + 48 + 20)
 #endif
 
 #endif  /*  __KERNEL__  */
