--- ./vanilla//drivers/mmc/core/sdio_cis.c	2010-02-23 17:38:51.000000000 +0200
+++ ./default//drivers/mmc/core/sdio_cis.c	2012-05-23 10:43:58.156935483 +0300
@@ -20,6 +20,8 @@
 #include <linux/mmc/sdio.h>
 #include <linux/mmc/sdio_func.h>
 
+#include <asm/mach-types.h>
+
 #include "sdio_cis.h"
 #include "sdio_ops.h"
 
@@ -140,7 +142,7 @@
 		return -EILSEQ;
 
 	vsn = func->card->cccr.sdio_vsn;
-	min_size = (vsn == SDIO_SDIO_REV_1_00) ? 28 : 42;
+	min_size = (vsn == SDIO_SDIO_REV_1_00) ? 28 : 34;
 
 	if (size < min_size || buf[0] != 1)
 		return -EINVAL;
@@ -242,8 +244,13 @@
 			break;
 
 		/* null entries have no link field or data */
-		if (tpl_code == 0x00)
-			continue;
+		if (tpl_code == 0x00) {
+			if (machine_is_msm8x55_svlte_surf() ||
+				machine_is_msm8x55_svlte_ffa())
+				break;
+			else
+				continue;
+		}
 
 		ret = mmc_io_rw_direct(card, 0, 0, ptr++, 0, &tpl_link);
 		if (ret)
