--- ./vanilla//drivers/staging/android/logger.h	2010-02-23 17:38:51.000000000 +0200
+++ ./default//drivers/staging/android/logger.h	2012-05-23 10:43:59.208935438 +0300
@@ -32,6 +32,7 @@
 
 #define LOGGER_LOG_RADIO	"log_radio"	/* radio-related messages */
 #define LOGGER_LOG_EVENTS	"log_events"	/* system/hardware events */
+#define LOGGER_LOG_SYSTEM	"log_system"	/* system/framework messages */
 #define LOGGER_LOG_MAIN		"log_main"	/* everything else */
 
 #define LOGGER_ENTRY_MAX_LEN		(4*1024)
