--- ./vanilla//drivers/video/msm/i2c_master_sw.h	1970-01-01 03:00:00.000000000 +0300
+++ ./default//drivers/video/msm/i2c_master_sw.h	2012-05-23 10:44:04.187935216 +0300
@@ -0,0 +1,8 @@
+#ifndef I2C_MASTER_SW_H
+#define I2C_MASTER_SW_H
+byte I2C_ReadByte(byte deviceID, byte offset);
+void I2C_WriteByte(byte deviceID, byte offset, byte value);
+byte I2C_ReadBlock(byte deviceID, byte offset, byte* buffer, word length);
+byte I2C_ReadSegmentBlock(byte deviceID, byte segment, byte offset, byte *buffer, word length);
+byte I2C_WriteBlock(byte deviceID, byte offset, byte* buffer, word length);
+#endif
