--- ./vanilla//include/linux/i2c/tsc2007.h	2010-02-23 17:38:51.000000000 +0200
+++ ./default//include/linux/i2c/tsc2007.h	2012-05-23 10:44:04.568935199 +0300
@@ -6,6 +6,11 @@
 struct tsc2007_platform_data {
 	u16	model;				/* 2007. */
 	u16	x_plate_ohms;
+	unsigned long irq_flags;
+	bool	invert_x;
+	bool	invert_y;
+	bool	invert_z1;
+	bool	invert_z2;
 
 	int	(*get_pendown_state)(void);
 	void	(*clear_penirq)(void);		/* If needed, clear 2nd level
