--- ./vanilla//drivers/serial/serial_core.c	2010-02-23 17:38:51.000000000 +0200
+++ ./default//drivers/serial/serial_core.c	2012-05-23 10:43:59.194935439 +0300
@@ -94,6 +94,9 @@
 	struct uart_state *state = tty->driver_data;
 	struct uart_port *port = state->uart_port;
 
+	if (port->ops->wake_peer)
+		port->ops->wake_peer(port);
+
 	if (!uart_circ_empty(&state->xmit) && state->xmit.buf &&
 	    !tty->stopped && !tty->hw_stopped)
 		port->ops->start_tx(port);
