--- ./vanilla//arch/arm/mm/dma-mapping.c	2010-02-23 17:38:51.000000000 +0200
+++ ./default//arch/arm/mm/dma-mapping.c	2012-05-23 10:43:54.074935665 +0300
@@ -544,8 +544,6 @@
 	void (*inner_op)(const void *, const void *);
 	void (*outer_op)(unsigned long, unsigned long);
 
-	BUG_ON(!virt_addr_valid(start) || !virt_addr_valid(start + size - 1));
-
 	switch (direction) {
 	case DMA_FROM_DEVICE:		/* invalidate only */
 		inner_op = dmac_inv_range;
@@ -564,7 +562,16 @@
 	}
 
 	inner_op(start, start + size);
+
+#ifdef CONFIG_OUTER_CACHE
+	/*
+	 * A page table walk would be required if the address isnt linearly
+	 * mapped. Simply BUG_ON for now.
+	 */
+	BUG_ON(!virt_addr_valid(start) || !virt_addr_valid(start + size - 1));
 	outer_op(__pa(start), __pa(start) + size);
+#endif
+
 }
 EXPORT_SYMBOL(dma_cache_maint);
 
