--- ./vanilla//drivers/usb/gadget/config.c	2010-02-23 17:38:51.000000000 +0200
+++ ./default//drivers/usb/gadget/config.c	2012-05-23 10:43:59.578935422 +0300
@@ -127,7 +127,7 @@
  * with identifiers (for interfaces, strings, endpoints, and more)
  * as needed by a given function instance.
  */
-struct usb_descriptor_header **__init
+struct usb_descriptor_header **
 usb_copy_descriptors(struct usb_descriptor_header **src)
 {
 	struct usb_descriptor_header **tmp;
@@ -174,7 +174,7 @@
  * intended use is to help remembering the endpoint descriptor to use
  * when enabling a given endpoint.
  */
-struct usb_endpoint_descriptor *__init
+struct usb_endpoint_descriptor *
 usb_find_endpoint(
 	struct usb_descriptor_header **src,
 	struct usb_descriptor_header **copy,
