--- ./vanilla//drivers/usb/gadget/gadget_chips.h	2010-02-23 17:38:51.000000000 +0200
+++ ./default//drivers/usb/gadget/gadget_chips.h	2012-05-23 10:43:59.584935422 +0300
@@ -169,6 +169,12 @@
 #define gadget_is_ci13xxx(g)	0
 #endif
 
+#ifdef CONFIG_USB_GADGET_MSM_72K
+#define	gadget_is_msm72k(g)	!strcmp("msm72k_udc", (g)->name)
+#else
+#define	gadget_is_msm72k(g)	0
+#endif
+
 // CONFIG_USB_GADGET_SX2
 // CONFIG_USB_GADGET_AU1X00
 // ...
@@ -247,6 +253,8 @@
 		return 0x24;
 	else if (gadget_is_r8a66597(gadget))
 		return 0x25;
+	else if (gadget_is_msm72k(gadget))
+		return 0x26;
 	return -ENOENT;
 }
 
