--- ./vanilla//drivers/usb/gadget/u_ether.c	2010-02-23 17:38:51.000000000 +0200
+++ ./default//drivers/usb/gadget/u_ether.c	2012-05-23 10:43:59.592935419 +0300
@@ -704,7 +704,7 @@
 MODULE_PARM_DESC(host_addr, "Host Ethernet Address");
 
 
-static u8 __init nibble(unsigned char c)
+static u8 nibble(unsigned char c)
 {
 	if (isdigit(c))
 		return c - '0';
@@ -714,7 +714,7 @@
 	return 0;
 }
 
-static int __init get_ether_addr(const char *str, u8 *dev_addr)
+static int get_ether_addr(const char *str, u8 *dev_addr)
 {
 	if (str) {
 		unsigned	i;
@@ -759,7 +759,7 @@
  *
  * Returns negative errno, or zero on success
  */
-int __init gether_setup(struct usb_gadget *g, u8 ethaddr[ETH_ALEN])
+int gether_setup(struct usb_gadget *g, u8 ethaddr[ETH_ALEN])
 {
 	struct eth_dev		*dev;
 	struct net_device	*net;
@@ -970,6 +970,8 @@
 	link->in = NULL;
 
 	usb_ep_disable(link->out_ep);
+	usb_ep_fifo_flush(link->in_ep);
+	usb_ep_fifo_flush(link->out_ep);
 	spin_lock(&dev->req_lock);
 	while (!list_empty(&dev->rx_reqs)) {
 		req = container_of(dev->rx_reqs.next,
