--- ./vanilla//drivers/gpu/drm/ati_pcigart.c	2010-02-23 17:38:51.000000000 +0200
+++ ./default//drivers/gpu/drm/ati_pcigart.c	2012-05-23 10:43:57.650935508 +0300
@@ -113,7 +113,7 @@
 
 		if (pci_set_dma_mask(dev->pdev, gart_info->table_mask)) {
 			DRM_ERROR("fail to set dma mask to 0x%Lx\n",
-				  gart_info->table_mask);
+				  (unsigned long long)gart_info->table_mask);
 			ret = 1;
 			goto done;
 		}
