--- ./vanilla//include/linux/mtd/partitions.h	2010-02-23 17:38:51.000000000 +0200
+++ ./default//include/linux/mtd/partitions.h	2012-05-23 10:44:04.599935198 +0300
@@ -51,6 +51,7 @@
 
 int add_mtd_partitions(struct mtd_info *, const struct mtd_partition *, int);
 int del_mtd_partitions(struct mtd_info *);
+void part_fill_badblockstats(struct mtd_info *mtd);
 
 /*
  * Functions dealing with the various ways of partitioning the space
