--- ./vanilla//drivers/usb/host/ehci-hcd.c	2010-02-23 17:38:51.000000000 +0200
+++ ./default//drivers/usb/host/ehci-hcd.c	2012-05-23 10:43:59.594935420 +0300
@@ -1131,6 +1131,11 @@
 #define	PLATFORM_DRIVER		ixp4xx_ehci_driver
 #endif
 
+#ifdef CONFIG_USB_EHCI_MSM
+#include "ehci-msm.c"
+#define PLATFORM_DRIVER		ehci_msm_driver
+#endif
+
 #ifdef CONFIG_USB_W90X900_EHCI
 #include "ehci-w90x900.c"
 #define	PLATFORM_DRIVER		ehci_hcd_w90x900_driver
